packages/pds/src/sequencer/index.ts

```diff
@@ -0,0 +1,181 @@
+    }
+
+    const getEvents = this.db.db
+      .selectFrom(seqQb.as('repo_seq'))
+      .leftJoin('repo_commit_history', (join) =>
+        join
+          .onRef('repo_commit_history.creator', '=', 'repo_seq.did')
+          .onRef('repo_commit_history.commit', '=', 'repo_seq.commit'),
+      )
+      .select([
+        'repo_seq.seq as seq',
+        'repo_seq.did as did',
+        'repo_seq.commit as commit',
+        'repo_seq.sequencedAt as sequencedAt',
+        'repo_commit_history.prev as prev',
+      ])
+      .orderBy('seq', 'asc')
+      .execute()
+
+    const getBlocks = this.db.db
+      .selectFrom(seqQb.as('repo_seq'))
+      .innerJoin('repo_commit_block', (join) =>
+        join
+          .onRef('repo_commit_block.creator', '=', 'repo_seq.did')
+          .onRef('repo_commit_block.commit', '=', 'repo_seq.commit'),
+      )
+      .innerJoin('ipld_block', (join) =>
+        join
+          .onRef('ipld_block.cid', '=', 'repo_commit_block.block')
+          .onRef('ipld_block.creator', '=', 'repo_commit_block.creator'),
+      )
+      .select([
+        'repo_seq.seq as seq',
+        'ipld_block.cid as cid',
+        'ipld_block.content as content',
+      ])
+      .execute()
+
+    const getBlobs = this.db.db
+      .selectFrom(seqQb.as('repo_seq'))
+      .innerJoin('repo_blob', (join) =>
+        join
+          .onRef('repo_blob.did', '=', 'repo_blob.did')
+          .onRef('repo_blob.commit', '=', 'repo_seq.commit'),
+      )
+      .select(['repo_seq.seq as seq', 'repo_blob.cid as cid'])
+      .execute()
+
+    const [events, blocks, blobs] = await Promise.all([
+      getEvents,
+      getBlocks,
+      getBlobs,
+    ])
+
+    const blocksBySeq = blocks.reduce((acc, cur) => {
+      acc[cur.seq] ??= new BlockMap()
+      acc[cur.seq].set(CID.parse(cur.cid), cur.content)
+      return acc
+    }, {} as Record<number, BlockMap>)
+
+    const blobsBySeq = blobs.reduce((acc, cur) => {
+      acc[cur.seq] ??= []
+      acc[cur.seq].push(cur.cid)
+      return acc
+    }, {} as Record<number, string[]>)
+
+    return Promise.all(
+      events.map(async (evt) => {
+        const commit = CID.parse(evt.commit)
+        const carSlice = await writeCar(commit, async (car) => {
+          const blocks = blocksBySeq[evt.seq]
+          if (blocks) {
+            for (const block of blocks.entries()) {
+              await car.put(block)
+            }
+          }
+        })
+        const blobs = blobsBySeq[evt.seq] || []
+        return {
+          seq: evt.seq,
+          time: evt.sequencedAt,
+          repo: evt.did,
+          commit: evt.commit,
+          prev: evt.prev || undefined,
+          blocks: carSlice,
+          blobs,
+        } as RepoAppendEvent
+      }),
+    )
+  }
+
+  async pollDb() {
+    const evts = await this.requestSeqRange({ earliestSeq: this.lastSeen })
+    if (evts.length > 0) {
+      this.emit('events', evts)
+      this.lastSeen = evts[evts.length - 1].seq
```

Comment:
Do we think this should be an invariant?
```js
sequencer.on('event', (evt) => {
  assert(evt.seq <= sequencer.lastSeen)
})
```

I.e. is the sequencer always first to see an event, and then its consumers get to see it?  If that seems right, we might consider swapping the order of these lines.

Reply:
ummm yeah i think that's right 👍